backend/src/apiserver/server/util.go

```diff
@@ -75,13 +79,34 @@ func DecompressPipelineTarball(compressedFile []byte) ([]byte, error) {
 	if err != nil {
 		return nil, util.NewInvalidInputErrorWithDetails(err, "Error extracting pipeline from the tarball file. Not a valid tarball file.")
 	}
+	//New behavior: searching for the "pipeline.yaml" file.
 	tarReader := tar.NewReader(gzipReader)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			tarReader = nil
+			break
+		}
+		if err != nil {
+			return nil, util.NewInvalidInputErrorWithDetails(err, "Error extracting pipeline from the tarball file. Not a valid tarball file.")
+		}
+		if isPipelineYamlFile(header.Name) {
+			//Found the pipeline file.
+			break
+		}
+	}
+	//Old behavior - taking the first file in the archive
+	if tarReader == nil {
+		gzipReader, err = gzip.NewReader(bytes.NewReader(compressedFile))
+		tarReader = tar.NewReader(gzipReader)
+	}
+
 	header, err := tarReader.Next()
-	if err != nil || header == nil {
+	if err != nil {
 		return nil, util.NewInvalidInputErrorWithDetails(err, "Error extracting pipeline from the tarball file. Not a valid tarball file.")
 	}
 	if !isYamlFile(header.Name) {
-		return nil, util.NewInvalidInputError("Error extracting pipeline from the tarball file. Expecting a YAML file inside the tarball. Got: %v", header.Name)
+		return nil, util.NewInvalidInputError("Error extracting pipeline from the tarball file. Expecting a pipeline.yaml file inside the tarball. Got: %v", header.Name)
```

Comment:
Isn't this just checking YAML format? Won't the previous error message be right then?

Reply:
This is fallback case which can only happen when there is no `pipeline.yaml` file in the archive. I want the error message to tell the user what they need to change to make the error go away - they need to make sure there is `pipeline.yaml` file inside the archive.

The completely pedantic error message would have been "The pipeline.yaml file wasn't found in archive and the *first* file in archive does not have .yaml extension.", but I think that the current message is better.